packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -395,6 +395,36 @@ abstract class Widget extends DiagnosticableTree {
   }
 }
 
+// A cache of inherited elements with an optional parent pointer.
+//
+// When looking up an inherited element, this will look through parent
+// pointers until the element is found or the end is reached. When an element
+// is found, it is cached at the first element where the search began.
+//
+// The intention of this cache is to speed up the initial build of widget
+// trees that contain a significant number of inherited widgets by deferring
+// expensive map allocations until they are needed.
+class _InheritedLookup {
+  _InheritedLookup? parent;
+  final HashMap<Type, InheritedElement> current = HashMap<Type, InheritedElement>();
```

Comment:
I kind of wonder if there's some way we can rework this so that there's only a single HashMap per depth of the tree, rather than having every element get one of these. It could live on the build owner and you could basically say "get me the inherited element for my current depth". You'd need to update it when an inherited element gets a new depth.

Does that make sense?

Reply:
Don't we have fewer hashmaps than per depth of the tree today? since we push down the hashmap created by an inherited element to all non-inherited elements

Reply:
I think you're right - I was thinking about cases where you might have a multi-child widget that has multiple sibling children that are inherited widgets, but in that case my suggestion doesn't even make sense.

Reply:
Also, I'm realizing now tht this will completely avoid copying the hash maps and my comment below doesn't make sense either. Sorry!

I'm trying to think if there's some way we can test this.

Reply:
WDYT of making this class public and just adding some unit tests for it?

That way we could have some tests that are more targetted and won't just break because someone refactors this and just starts getting errors about some weird inherited widget.

Reply:
Sure, can do!

Reply:
done